controllers/openstackipset_controller.go

```diff
@@ -0,0 +1,397 @@
+		delete(instance.Status.Hosts, deletedHost)
+	}
+
+	//
+	// create hostnames
+	//
+	_, err = r.createNewHostnames(
+		instance,
+		cond,
+		instance.Spec.HostCount-len(instance.Status.Hosts),
+	)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	//
+	// get OSNetCfg object
+	//
+	osnetcfg := &ospdirectorv1beta1.OpenStackNetConfig{}
+	err = r.Client.Get(context.TODO(), types.NamespacedName{
+		Name:      strings.ToLower(instance.Labels[openstacknetconfig.OpenStackNetConfigReconcileLabel]),
+		Namespace: instance.Namespace},
+		osnetcfg)
+	if err != nil {
+		cond.Message = fmt.Sprintf("Failed to get %s %s ", osnetcfg.Kind, osnetcfg.Name)
+		cond.Reason = ospdirectorv1beta1.NetConfigCondReasonnError
+		cond.Type = ospdirectorv1beta1.CommonCondTypeError
+		err = common.WrapErrorForObject(cond.Message, instance, err)
+
+		return ctrl.Result{}, err
+	}
+
+	//
+	// Wait for IPs created on all configured networks
+	//
+	for hostname, hostStatus := range instance.Status.Hosts {
+		err = openstacknetconfig.WaitOnIPsCreated(
+			r,
+			instance,
+			cond,
+			osnetcfg,
+			instance.Spec.Networks,
+			hostname,
+			&hostStatus,
+		)
+		if err != nil {
+			return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+		}
+
+		//
+		// check if owning object is OSBMS
+		//
+		osBms := &ospdirectorv1beta1.OpenStackBaremetalSet{}
+		err = r.Client.Get(context.TODO(), types.NamespacedName{
+			Name:      instance.Labels[common.OwnerNameLabelSelector],
+			Namespace: instance.Namespace},
+			osBms)
+		if err != nil && !k8s_errors.IsNotFound(err) {
+			cond.Message = fmt.Sprintf("Failed to get %s %s ", osBms.Kind, osBms.Name)
+			cond.Reason = ospdirectorv1beta1.BaremetalSetCondReasonError
+			cond.Type = ospdirectorv1beta1.CommonCondTypeError
+			err = common.WrapErrorForObject(cond.Message, instance, err)
+
+			return ctrl.Result{}, err
+		}
+
+		// Can not set (like in vmset, osclient, osctlplane) HostRef for BMS to hostname as it references the used BMH host
+		// if owning object is not osbms
+		if k8s_errors.IsNotFound(err) {
+			hostStatus.HostRef = hostname
+		} else {
+
+			// Get openshift-machine-api BaremetalHosts with
+			baremetalHostsList := &metal3v1alpha1.BareMetalHostList{}
+
+			labelSelector := map[string]string{
+				common.OSPHostnameLabelSelector: hostname,
+			}
+
+			listOpts := []client.ListOption{
+				client.InNamespace("openshift-machine-api"),
+				client.MatchingLabels(labelSelector),
+			}
+
+			err := r.GetClient().List(context.TODO(), baremetalHostsList, listOpts...)
+			if err != nil && !k8s_errors.IsNotFound(err) {
+				cond.Message = "Failed to get list of all BareMetalHost(s)"
+				cond.Reason = ospdirectorv1beta1.BaremetalHostCondReasonListError
+				cond.Type = ospdirectorv1beta1.ConditionType(ospdirectorv1beta1.BaremetalSetCondTypeError)
+
+				return reconcile.Result{}, err
+			}
+
+			for _, bmh := range baremetalHostsList.Items {
+				hostStatus.HostRef = bmh.Name
+			}
```

Comment:
Are we only ever expecting one item in the list?  Just wondering if things would get wonky if somehow this list had more than one element.

Reply:
yes, we expect that there is a 1:1 mapping for osp hostname and bmh. if we get multiple BMHs with that above labelselector, there was an error in the bmset provisioning of bmh hosts [1]. 

[1] https://github.com/openstack-k8s-operators/osp-director-operator/blob/master/controllers/openstackbaremetalset_controller.go#L1025